User Authentication

NabeelMinhas opened the pull request:
Complete user authentication using the devise gem

NabeelMinhas committed:
Complete user authentication using devise gem

Hassan01-dev commented on Gemfile:
```diff
@@ -71,3 +71,5 @@ group :test do
 
 end
+
+gem "devise", "~> 4.9"
```
Add a small comment about why we are using this gem like
```suggestion
gem "devise", "~> 4.9" # User Authorization
```
or something else. Also move this gem above all the groups in file

Hassan01-dev commented on app/models/user.rb:
```diff
@@ -0,0 +1,18 @@
+  devise :database_authenticatable, :registerable,
+         :recoverable, :rememberable, :validatable, :confirmable
+
+  USER_CATEGORY = [
+    ["manager", "Manager"],
+    ["developer", "Developer"],
+    ["qa", "QA"],
+  ]
```
We can use rails method `.titalize` to convert "manager" to "Manager".
If we still need this then move this into helper instead of model

NabeelMinhas replied:
I am moving this to helper

Hassan01-dev commented on app/models/user.rb:
```diff
@@ -0,0 +1,18 @@
+    ["qa", "QA"],
+  ]
+
+  enum user_category: {
+    manager: 0,
+    developer: 1,
+    qa: 2
+  }
```
We can make this more readable like this
```suggestion
  USER_CATEGORIES = { manager: 0, developer: 1, qa: 2 }.freeze
  enum user_category: USER_CATEGORIES
```

NabeelMinhas replied:
I think by doing this. We are not improving the readability but better memory consumption

Hassan01-dev commented on app/views/devise/registrations/edit.html.erb:
```diff
@@ -0,0 +1,53 @@
+
+  <div class="field">
+    <%= f.label :user_category %><br />
+    <%= f.select :user_category, User::USER_CATEGORY, autofocus: true, autocomplete: "user_category" %>
```
Instead of getting these from model. We should use helper method `User::USER_CATEGORY`
Also don't use autocomplete for dropdown or select field type, instead use the value which is saved in the DB as this is the edit form.
If its working fine then leave it

NabeelMinhas replied:
I am using the helper method instead of User::USER_CATEGORY.

Hassan01-dev commented on app/views/devise/passwords/edit.html.erb:
```diff
@@ -0,0 +1,25 @@
+<h2>Change your password</h2>
```
If we didn't make any change in the devise view we can remove them.
But its upto you how you want to manage this

NabeelMinhas replied:
sure

Hassan01-dev commented on app/views/devise/registrations/new.html.erb:
```diff
@@ -0,0 +1,39 @@
+
+  <div class="field">
+    <%= f.label :user_category %><br />
+    <%= f.select :user_category, User::USER_CATEGORY, {} %>
```
Same as above

NabeelMinhas replied:
okay

Hassan01-dev commented on app/views/devise/registrations/edit.html.erb:
```diff
@@ -0,0 +1,53 @@
+
+<h3>Cancel my account</h3>
+
+<div>Unhappy? <%= button_to "Cancel my account", registration_path(resource_name), data: { confirm: "Are you sure?", turbo_confirm: "Are you sure?" }, method: :delete %></div>
```
We only need one confirm block, which is turbo_confirm. Remove the extra one but if you don't want to then leave it

NabeelMinhas replied:
I am leaving it

Hassan01-dev commented on config/environments/development.rb:
```diff
@@ -36,6 +36,8 @@
   config.active_storage.service = :local
 
+  config.action_mailer.default_url_options = { host: 'localhost', port: 3000 }
+
```
We might also need to add letter_opener on development or add smtp configuration for sending mails

NabeelMinhas replied:
sure

Hassan01-dev commented on config/routes.rb:
```diff
@@ -1,4 +1,5 @@
 Rails.application.routes.draw do
+  devise_for :users
   # Define your application routes per the DSL in https://guides.rubyonrails.org/routing.html
 
```
Also add a root_path, which is required for devise to redirect in case user is not authenticated.

NabeelMinhas replied:
okay

Hassan01-dev commented on db/migrate/20230916101807_devise_create_users.rb:
```diff
@@ -0,0 +1,46 @@
+      t.string   :confirmation_token
+      t.datetime :confirmed_at
+      t.datetime :confirmation_sent_at
+      t.string   :unconfirmed_email # Only if using reconfirmable
```
Are we using `reconfirmable`? If not then you can comment this line.

NabeelMinhas replied:
I want re-confirmable because sometimes we need to send the confirmation token again

Hassan01-dev reviewed (changes requested):
Few Minor changes

NabeelMinhas committed:
Address feedback

- Move helper method in views helper instead of model
- Add letter opener in development environment
- Remove extra files

Hassan01-dev commented on app/helpers/application_helper.rb:
```diff
@@ -1,2 +1,9 @@
 module ApplicationHelper
+  def user_category_options
+    [
+      ["manager", "Manager"],
+      ["developer", "Developer"],
+      ["qa", "QA"],
+    ]
```
```suggestion
    User::USER_CATEGORIES.keys.map { |x| [x.to_s, x.to_s.titleize] }
```

That way you don't need to update the array every time you make a change in the `USER_CATEGORIES`

Hassan01-dev commented on app/models/user.rb:
```diff
@@ -0,0 +1,10 @@
+class User < ApplicationRecord
+  # Include default devise modules. Others available are:
+  # :lockable, :timeoutable, :trackable and :omniauthable
```
Remove these comments

Hassan01-dev reviewed (approved):
LGTM, just a suggestion. You can implement that in some other PR. For now lets merge this one.
